src/daft-recordbatch/src/ops/window.rs

```diff
@@ -108,4 +111,74 @@ impl RecordBatch {
         // Union the original data with the row number column
         self.union(&row_number_batch)
     }
+
+    pub fn window_rank(&self, name: String, order_by: &[ExprRef], dense: bool) -> DaftResult<Self> {
+        // Create rank numbers for pre-sorted partition
+        let mut rank_numbers = vec![0u64; self.len()];
+
+        if self.is_empty() {
+            // Empty partition case - no work needed
+            let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+            let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+            return self.union(&rank_batch);
+        }
+
+        // Single row case - always rank 1
+        if self.len() == 1 {
+            rank_numbers[0] = 1;
+            let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+            let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+            return self.union(&rank_batch);
+        }
+
+        // For partitions with multiple rows, compute ranks
+        let mut cur_rank = 1; // Start at rank 1
+        let mut next_rank = 1;
+
+        // Get the order_by columns
+        let order_by_table = self.eval_expression_list(order_by)?;
+
+        // Create a comparator for checking equality between rows
+        let comparator: Box<dyn Fn(usize, usize) -> bool + Send + Sync> =
+            build_multi_array_is_equal(
+                order_by_table.columns.as_slice(),
+                order_by_table.columns.as_slice(),
+                &vec![true; order_by_table.columns.len()],
+                &vec![true; order_by_table.columns.len()],
+            )?;
+
+        // First row always has rank 1
+        rank_numbers[0] = 1;
+
+        // Compute ranks for remaining rows
+        for (i, rank) in rank_numbers.iter_mut().enumerate().skip(1) {
+            // Always increment next_rank for regular rank()
+            if !dense {
+                next_rank += 1;
+            }
+
+            // Check if the current row has the same values as the previous row
+            let is_equal = comparator(i - 1, i);
+
+            if !is_equal {
+                // Different value, update rank
+                if dense {
+                    // For dense_rank, just increment by 1
+                    cur_rank += 1;
+                } else {
+                    // For rank(), use the next_rank which accounts for ties
+                    cur_rank = next_rank;
+                }
+            }
+
+            *rank = cur_rank;
+        }
```

Comment:
This is a bit of a nit, but instead of creating the `rank_numbers` vec and doing an iter_mut() on it, you could just do something like 
```rust
std::iter::once(1)
    .chain((1..self.len()).map(|i| {
        // the logic here
    }))
    .collect()
```
might be cleaner?